benchmark/src/benchmark_types.ts

```diff
@@ -0,0 +1,22 @@
+export enum BenchmarkKind {
+    Memory = "memory",
+}
+
+export interface BenchmarkResult {
+    kind: BenchmarkKind;
+}
+
+export type BenchmarkFunction = () => void;
+
+export interface MemoryBenchmarkResult extends BenchmarkResult {
+    kind: BenchmarkKind.Memory;
+    benchmarkName: string;
+    preExecMemoryUsage: number;
+    postExecMemoryUsage: number;
+    memoryUsedForExec: number;
+    memoryAfterGC: number;
+}
+
+export function isMemoryBenchmarkResult(result: BenchmarkResult): result is MemoryBenchmarkResult {
+    return result.kind == BenchmarkKind.Memory;
+}
```

Comment:
Sum types are generally easier to use than a common interface with manual type guards

```suggestion
export type BenchmarkFunction = () => void;

export enum BenchmarkKind {
    Memory = "memory",
}

type BenchmarkResult = MemoryBenchmarkResult;

export interface MemoryBenchmarkResult {
    kind: BenchmarkKind.Memory;
    benchmarkName: string;
    preExecMemoryUsage: number;
    postExecMemoryUsage: number;
    memoryUsedForExec: number;
    memoryAfterGC: number;
}
```